src/internal/pjs/server_test.go

```diff
@@ -71,6 +74,114 @@ func TestInspectJob(t *testing.T) {
 	})
 }
 
+func TestDo(t *testing.T) {
+	c, fc := setupTest(t)
+	ctx := pctx.TestContext(t)
+	programFileset := createFileSet(t, fc, map[string][]byte{
+		"file": []byte(`!#/bin/bash; ls /input/;`),
+	})
+	inputFileset := createFileSet(t, fc, map[string][]byte{
+		"a.txt": []byte("dummy input"),
+	})
+	in := &pjs.CreateJobRequest{
+		Program: programFileset,
+		Input:   []string{inputFileset},
+	}
+	out, err := do(t, ctx, c, in, func(inputs []string) ([]string, error) {
+		// todo(muyang)
+		// for now we do nothing here, simply return the input filesets
+
+		return inputs, nil
+	})
+	require.NoError(t, err)
+	require.Equal(t, in.Input, out.Output)
+}
+
+// do does work through PJS.
+func do(t *testing.T, ctx context.Context, s pjs.APIClient, in *pjs.CreateJobRequest, fn func(inputs []string) ([]string, error)) (*pjs.JobInfo_Success, error) {
```

Comment:
[nit]: I dislike the name 'do' its too generic. Rename this with the *what* it is doing. 
For example, here do creates jobs from in and processes them until they are done.
You could call this doJobWork() or runJob() for example.